codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerUpgradeCodec.java

```diff
@@ -151,8 +148,23 @@ public boolean prepareUpgradeResponse(ChannelHandlerContext ctx, FullHttpRequest
 
     @Override
     public void upgradeTo(final ChannelHandlerContext ctx, FullHttpRequest upgradeRequest) {
-        // Add the HTTP/2 connection handler to the pipeline immediately following the current handler.
-        ctx.pipeline().addAfter(ctx.name(), handlerName, upgradeToHandler);
+        try {
+            // Add the HTTP/2 connection handler to the pipeline immediately following the current handler.
+            ctx.pipeline().addAfter(ctx.name(), handlerName, connectionHandler);
+            connectionHandler.onHttpServerUpgrade(settings);
+
+        } catch (Http2Exception e) {
+            ctx.fireExceptionCaught(e);
+            ctx.close();
+            return;
+        }
+
+        if (handlers != null) {
+            final String name = ctx.pipeline().context(connectionHandler).name();
+            for (int i = handlers.length - 1; i >= 0; i--) {
+                ctx.pipeline().addAfter(name, null, handlers[i]);
```

Comment:
can upgradeTo be called multiple times?  If not, does the codec live possibly a long time?  If both of those things, then should handlers be null'ed out after this call?